### PR TITLE
Documentation and new types for the internal package

### PR DESCRIPTION
**Pull request by msehnout**
TODO
- [x] Implement MarshalJSON, UnmarshalJSON for ImageType
- [x] Implement MarshalJSON, UnmarshalJSON for Distribution
- [x] Implement MarshalJSON, UnmarshalJSON for UploadTarget

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -22,3 +22,35 @@ or using a browser: `http://localhost:9090`
 
 Please refer to the [lorax-composer](https://github.com/weldr/lorax)'s documenation as osbuild-composer is a drop-in replacement.
 
+## High-level overview
+
+![overview](docs/osbuild-composer.svg)
+
+### Frontends
+
+`osbuild-composer` is meant to be used with 3 different front-ends. The primary one, which is meant for general use, is cockpit-composer. It is part of the Cockpit project and unless you have a strong reason not to use it, you should use it. `composer-cli` is a command line tool that can be used for non-interactive usage. Finally, RCM client (TBD) is a special-purpose client that is not meant for general usage.
```
Comment (bcl) [inline README.md]: `composer-cli` is *not* for non-interactive usage. It is for humans and makes no guarantees about its output. All non-interactive usage outside of tests that are tightly coupled to the code repository should be using the API directly.

File: README.md [hunk 2]
```diff
@@ -22,3 +22,35 @@ or using a browser: `http://localhost:9090`
...
+
+### Compose
+* Compose is what the user submits over one of the frontends
+* It contains of one or more **image builds**
+* It contains zero or more **upload actions**
+
+### Image build
+* The resulting *image* has a *type*: https://github.com/osbuild/osbuild-composer/blob/master/internal/distro/fedora30/distro.go#L19
```
Comment (bcl) [inline README.md]: That link doesn't tell me anything about supported types. Maybe it should reference `composer-cli compose types` instead?

File: internal/common/types.go [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+		*arch = S390x
+	default:
+		return &ParseCustomTypeError{archString + " is not a valid architecture"}
+	}
+	return nil
+}
+
+type ImageType int
+
+const (
```
Comment (bcl) [inline internal/common/types.go]: I think this and Distribution should also have unmarshal functions, depending on int values for the inputs is going to be fragile and difficult to keep synchrnized with the front ends when they are extended.

File: internal/crypt/crypt_impl.go [hunk 4]
```diff
@@ -1,3 +1,5 @@
+// +build !macos
```
Comment (bcl) [inline internal/crypt/crypt_impl.go]: What is `+build !macos`?

File: internal/crypt/crypt_impl_macos.go [hunk 5]
```diff
@@ -0,0 +1,7 @@
+// +build macos
+
+package crypt
+
+func crypt(pass, salt string) (string, error) {
+	return "", nil
```
Comment (bcl) [inline internal/crypt/crypt_impl_macos.go]: Why is this here? This seems like the kind of thing that can come back to haunt you in the future...

File: README.md (hunk 1 above)
Comment (msehnout) [reply README.md]: thanks, I didn't know that

File: README.md (hunk 2 above)
Comment (msehnout) [reply README.md]: I think about this README file as a file for developers, but it is worth mentioning.

File: internal/common/types.go (hunk 3 above)
Comment (msehnout) [reply internal/common/types.go]: Sure, I'll implement Unmarshal as well. Would you suggest to use something else than `int`?

File: internal/crypt/crypt_impl.go (hunk 4 above)
Comment (msehnout) [reply internal/crypt/crypt_impl.go]: Conditional compilation for all non-macos systems.

File: internal/crypt/crypt_impl_macos.go (hunk 5 above)
Comment (msehnout) [reply internal/crypt/crypt_impl_macos.go]: yeah, @ondrejbudai advised me to use panic instead, which is very, very resonable :)

Comment (msehnout) [reply internal/crypt/crypt_impl_macos.go]: And it's here to make compilation on macos possible, because glibc does not support macos so the implementation in `crypt_impl` cannot be compiled...

File: README.md [hunk 6]
```diff
@@ -22,3 +22,36 @@ or using a browser: `http://localhost:9090`
 
 Please refer to the [lorax-composer](https://github.com/weldr/lorax)'s documenation as osbuild-composer is a drop-in replacement.
 
+## High-level overview
+
+![overview](docs/osbuild-composer.svg)
+
+### Frontends
+
+`osbuild-composer` is meant to be used with 3 different front-ends. The primary one, which is meant for general use, is cockpit-composer. It is part of the Cockpit project and unless you have a strong reason not to use it, you should use it. `composer-cli` is a command line tool that can be used with `osbuild-composer`.
```
Comment (teg) [inline README.md]: Maybe change to "`osbuild-composer` has two frontends:" (rather than 3, no need to talk about RCM yet).

File: internal/common/types.go [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+type ParseCustomTypeError struct {
+	reason string
+}
+
+func (err *ParseCustomTypeError) Error() string {
+	return err.reason
+}
+
+// Architecture represents one of the supported CPU architectures available for images
+// produced by osbuild-composer. I decided to represent it as an integer because if it
```
Comment (teg) [inline internal/common/types.go]: Use the passive voice in comments: "It is represented as an integer because...". The same for the rest of the comments. This doesn't make a difference to the meaning of course, just keeps the style consistent.

File: internal/common/types.go [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+// was a string it would unmarshal from JSON just fine even in case that the architecture
+// was unknown (yeah, thanks Go for doing the same mistakes C did almost 50 years ago)
```
Comment (teg) [inline internal/common/types.go]: I don't disagree, but drop the comments in the parenthesis to make the tone of the documentation more neutral.

File: internal/common/types.go [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+//
+// I considered some alternatives like a struct with private fields, but struct cannot
+// be const, so that does not help either. Eventually I think this is the "get s**t done"
+// solution and next time please use a language with decent type system.
```
Comment (teg) [inline internal/common/types.go]: This comment would be better suited to the commit message ("why I did what I did", as opposed to "how it works"). And comments about Go would be better suited to Twitter ;)

File: internal/store/store.go [hunk 10]
```diff
@@ -489,6 +489,11 @@ func (s *Store) PushCompose(composeID uuid.UUID, bp *blueprint.Blueprint, checks
 	return nil
 }
 
+func (s *Store) PushRcmCompose(composeRequest ComposeRequest) error {
```
Comment (teg) [inline internal/store/store.go]: This looks like it belongs in a different commit.

File: README.md (hunk 2 above)
Comment (bcl) [reply README.md]: Even as a developer that line didn't help. I see that variable is populated later with the types but it's not immediately clear when looking at that line.

File: internal/common/types.go (hunk 3 above)
Comment (bcl) [reply internal/common/types.go]: I think using int constants internally is fine.

File: internal/common/types.go (hunk 7 above)
Comment (msehnout) [reply internal/common/types.go]: Sure. It's just that all the grammar-checking tools nowadays complain about overuse of passive voice, that I'm trying to avoid it from time to time.

File: internal/common/types.go (hunk 9 above)
Comment (msehnout) [reply internal/common/types.go]: Hehe, I don't use Twitter or any other social network, but I'll drop it if you insist!

File: internal/store/store.go (hunk 10 above)
Comment (msehnout) [reply internal/store/store.go]: You are right.

File: README.md (hunk 2 above)
Comment (msehnout) [reply README.md]: Sure, but I still think it is a pointer in the right direction. There is a whole section about "image type" below this one.

File: internal/common/types.go [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+
+const (
+	Fedora30 Distribution = iota
+	RHEL82
+)
+
+type UploadTarget int
+
+const (
+	S3           UploadTarget = iota
```
[the remark this answers is not in the document]
Comment (msehnout) [reply internal/common/types.go]: done

File: README.md [hunk 12]
```diff
@@ -22,3 +22,35 @@ or using a browser: `http://localhost:9090`
...
+![overview](docs/osbuild-composer.svg)
+
+### Frontends
+
+`osbuild-composer` is meant to be used with 3 different front-ends. The primary one, which is meant for general use, is cockpit-composer. It is part of the Cockpit project and unless you have a strong reason not to use it, you should use it. `composer-cli` is a command line tool that can be used for non-interactive usage. Finally, RCM client (TBD) is a special-purpose client that is not meant for general usage.
+
+### Compose
+* Compose is what the user submits over one of the frontends
+* It contains of one or more **image builds**
+* It contains zero or more **upload actions**
```
[the remark this answers is not in the document]
Comment (msehnout) [reply README.md]: done

File: README.md [hunk 13]
```diff
@@ -22,3 +22,35 @@ or using a browser: `http://localhost:9090`
...
+
+### Image build
+* The resulting *image* has a *type*: https://github.com/osbuild/osbuild-composer/blob/master/internal/distro/fedora30/distro.go#L19
+* Running build in osbuild-composer is referred to as a "job" (internal terminology, not related to end-user experience)
+
+### Job
+* What composer submits to a worker
+* Is a unit of work performed by `osbuild` (internally it is a single execution of `osbuild`)
+* Consists of **one** image build and **one or more** Upload actions
```
[the remark this answers is not in the document]
Comment (msehnout) [reply README.md]: fixed

File: internal/common/types.go [hunk 14]
```diff
@@ -0,0 +1,212 @@
+package store
```
Comment (ondrejbudai) [inline internal/common/types.go]: Is this right? I'm not sure if this belongs to the `store` package.

Comment (msehnout) [reply internal/common/types.go]: Thanks for catching this! I guess `package common` was in the commit I removed...
